File: ts/src/framing.ts
import * as net from 'net';

/**
 * BufferedReader wraps a net.Socket and provides an async `readExactly(n)` 
 * method. It uses a single continuous 'data' listener to buffer all incoming
 * data, ensuring frames are never interleaved or lost.
 */
export class BufferedReader {
  private buf = Buffer.alloc(0);
  private waiters: Array<(buf: Buffer) => void> = [];
  private closed = false;

  constructor(private socket: net.Socket) {
    socket.on('data', (chunk: Buffer) => {
      this.buf = Buffer.concat([this.buf, chunk]);
      this._drainWaiters();
    });

    socket.once('close', () => {
      this.closed = true;
      // Reject all pending waiters
      for (const w of this.waiters) {
        w(Buffer.alloc(0));
      }
      this.waiters = [];
    });

    socket.once('error', () => {
      this.closed = true;
      for (const w of this.waiters) {
        w(Buffer.alloc(0));
      }
      this.waiters = [];
    });
  }

  private _drainWaiters(): void {
    while (this.waiters.length > 0) {
      // Peek at the first waiter's needed size (we can't know it here without
      // wrapping, so just wake every waiter and let them re-check)
      const waiter = this.waiters[0];
      waiter(Buffer.alloc(0)); // signal "data available"
      this.waiters.shift();
    }
  }

  async readExactly(n: number): Promise<Buffer> {
    while (this.buf.length < n) {
      if (this.closed) {
        throw new Error('Socket closed');
      }
      await new Promise<void>(resolve => {
        this.waiters.push(() => resolve());
      });
    }
    const result = this.buf.subarray(0, n);
    this.buf = this.buf.subarray(n);
    return result;
  }

  /**
   * Read one Callwire frame: [4-byte big-endian uint32 length][N bytes payload]
   */
  async readFrame(): Promise<Buffer> {
    const header = await this.readExactly(4);
    const len = header.readUInt32BE(0);
    if (len === 0) return Buffer.alloc(0);
    return this.readExactly(len);
  }
}

/**
 * Writes one Callwire frame to a socket.
 * Frame format: [4-byte big-endian uint32 length][N bytes msgpack payload]
 */
export function writeFrame(socket: net.Socket, payload: Buffer): boolean {
  const header = Buffer.allocUnsafe(4);
  header.writeUInt32BE(payload.length, 0);
  return socket.write(Buffer.concat([header, payload]));
}

File: c/src/framing.c
#include "callwire.h"
#include <stdlib.h>
#include <string.h>
#include <sys/socket.h>
#include <arpa/inet.h>
#include <unistd.h>

/* Read a frame from socket: [4-byte big-endian length][payload]. */
int callwire_framing_read(int sock, uint8_t **payload_out, size_t *len_out) {
    uint8_t header[4];
    ssize_t n = recv(sock, header, 4, 0);
    if (n != 4) {
        return -1; /* error or disconnect */
    }

    /* Parse big-endian uint32. */
    uint32_t len = ((uint32_t)header[0] << 24) |
                   ((uint32_t)header[1] << 16) |
                   ((uint32_t)header[2] << 8) |
                   ((uint32_t)header[3]);

    if (len > 16 * 1024 * 1024) {
        return -1; /* payload too large (16MB limit) */
    }

    uint8_t *payload = malloc(len);
    if (!payload) {
        return -1;
    }

    size_t read_so_far = 0;
    while (read_so_far < len) {
        ssize_t n = recv(sock, payload + read_so_far, len - read_so_far, 0);
        if (n <= 0) {
            free(payload);
            return -1;
        }
        read_so_far += n;
    }

    *payload_out = payload;
    *len_out = len;
    return 0;
}

/* Write a frame to socket. */
int callwire_framing_write(int sock, const uint8_t *payload, size_t len) {
    uint8_t header[4];
    header[0] = (len >> 24) & 0xFF;
    header[1] = (len >> 16) & 0xFF;
    header[2] = (len >> 8) & 0xFF;
    header[3] = len & 0xFF;

    if (send(sock, header, 4, 0) != 4) {
        return -1;
    }
    if (send(sock, payload, len, 0) != (ssize_t)len) {
        return -1;
    }
    return 0;
}
